docs/content/aws-cloudformation-hooks.md

```diff
@@ -0,0 +1,479 @@
+deny[msg] {
+    bucket_create_or_update
+    
+    input.resource.properties.AccessControl != "Private"
+
+    msg := sprintf("S3 Bucket %s 'AccessControl' attribute value must be 'Private'", [input.resource.id])
+}
+```
+
+This rule would work fine as long as there _is_ an `AccessControl` attribute present in the input resource, but would
+fail (i.e. not evaluate) as soon as the property was missing, leading to the resource being allowed! Using helper rules
+and negation is a good way to work with data that might or might not be present, and results in more readable policies
+too.
+
+## 5. Policy Enforcement Testing 
+
+With the above policy loaded into OPA, we may proceed to try it out. Let's deploy the minimal S3 Bucket from the 
+previous template example. Save the below minimal template to a file called `s3bucket.yaml`:
+
+```yaml
+Resources:
+  ExampleS3Bucket:
+    Type: AWS::S3::Bucket
+```
+
+Since our S3 bucket doesn't have an `AccessControl` attribute, it should be denied by the hook. We 
+deploy a template by creating a **stack**:
+
+```shell
+aws cloudformation create-stack --stack-name cfn-s3 --template-body file://s3bucket.yaml
+```
+
+The output of the above command will simply be a confirmation that the stack was deployed. It won't tell us whether the
+deployment was successful or not. In order to know that, we'll need to check the stack events:
+
+```shell
+aws cloudformation describe-stack-events --stack-name cfn-s3
+```
+
+The output of the above command will be a list of all events associated with the `cfn-s3` stack. Among the events, you
+should now find an item describing that the hook denied the request, and its reason for doing so:
+
+```json
+{
+  "StackEvents": [
+    {
+      "StackId": "arn:aws:cloudformation:eu-north-1:55523455647:stack/cfn-s3/4f605f70-b1ca-12ec-b4d8-0a63e869dfee",
+      "EventId": "ExampleS3Bucket-c243efd6-bfe7-3f10-8304-a0e40fe5f6f4",
+      "StackName": "cfn-s3",
+      "LogicalResourceId": "ExampleS3Bucket",
+      "PhysicalResourceId": "",
+      "ResourceType": "AWS::S3::Bucket",
+      "Timestamp": "2022-03-31T08:41:15.946000+00:00",
+      "ResourceStatus": "CREATE_IN_PROGRESS",
+      "HookType": "Styra::OPA::Hook",
+      "HookStatus": "HOOK_COMPLETE_FAILED",
+      "HookStatusReason": "Hook failed with message: S3 Bucket ExampleS3Bucket 'AccessControl' attribute value must be 'Private'",
+      "HookInvocationPoint": "PRE_PROVISION",
+      "HookFailureMode": "FAIL"
+    }
+  ]
+}
+```
+
+Congratulations! You've just successfully enforced your first CloudFormation Hook policy using OPA. Let's update the 
+template so that it passes our policy requirement:
+
+**s3bucket.yaml**
+```yaml
+Resources:
+  ExampleS3Bucket:
+    Type: AWS::S3::Bucket
+    Properties:
+      AccessControl: Private
+```
+
+Even though our stack did not create an S3 bucket (as the change got rolled back), the **stack** still exists. 
+In order to try again, we'll first need to delete the existing stack:
```

Comment:
It's fine for the purpose of this tutorial, but in real life,  could the be situations where deleting the stack is not an option? Can you go forward in some way anyways?

Reply:
Sure! You could just create a new stack named differently. There's also an `update-stack` command for deploying changes, but it only works once a stack has been successfully deployed. I can add a note on that.

Reply:
Done.